### PR TITLE
http on windows

### PR DESCRIPTION
**Pull request by 9cvele3**
Related to triton-inference-server/server#3130

PR for third_party: https://github.com/triton-inference-server/third_party/pull/9

### CODE REVIEW

File: src/servers/main.cc [hunk 1]
```diff
@@ -702,6 +707,11 @@ StartEndpoints(
     const std::shared_ptr<nvidia::inferenceserver::SharedMemoryManager>&
         shm_manager)
 {
+#ifdef _WIN32
+  WSADATA wsaData;
+  WSAStartup(MAKEWORD(2,2), &wsaData);
```
Comment (GuanLuo) [inline src/servers/main.cc:712]: Should check return value in case of error.

File: src/servers/http_server.cc [hunk 2]
```diff
@@ -24,6 +24,7 @@
 // (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+#define NOMINMAX
```
Comment (GuanLuo) [inline src/servers/http_server.cc:27]: Should this be surrounded by `ifdef _WIN32` like what is in `main.cc`?

**Review by GuanLuo — CHANGES_REQUESTED** [review]
And if you haven't done so, we need a signed CLA for external contribution https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla

**Comment by deadeyegoodwin** [comment]
Thanks for this contribution. We will work on integrating and testing it in our CI.

**Comment by 9cvele3** [comment]
> Thanks for this contribution. We will work on integrating and testing it in our CI.

OK, just keep in mind that there is also a PR for third_party (actually, most of the changes are there :smile:)
https://github.com/triton-inference-server/third_party/pull/9

**Review by GuanLuo — APPROVED** [review]
Looks good to me, the PR will need to wait until we add tests as mentioned

**Comment by deadeyegoodwin** [comment]
Did you build on Linux? I am seeing the following error:

```
/workspace/src/servers/http_server.cc: In member function ‘TRITONSERVER_Error* nvidia::inferenceserver::HTTPServer::Start()’:
/workspace/src/servers/http_server.cc:76:48: error: cannot convert ‘intptr_t*’ {aka ‘long int*’} to ‘int*’
   76 |     evutil_socketpair(AF_UNIX, SOCK_STREAM, 0, (intptr_t*)fds_);
      |                                                ^~~~~~~~~~~~~~~
      |                                                |
      |                                                intptr_t* {aka long int*}
In file included from /tmp/tritonbuild/tritonserver/build/third-party/libevent/include/event2/buffer.h:89,
                 from /workspace/build/server/../../src/servers/data_compressor.h:35,
                 from /workspace/build/server/../../src/servers/http_server.h:38,
                 from /workspace/src/servers/http_server.cc:31:
/tmp/tritonbuild/tritonserver/build/third-party/libevent/include/event2/util.h:392:70: note:   initializing argument 4 of ‘int evutil_socketpair(int, int, int, int*)’
  392 | int evutil_socketpair(int d, int type, int protocol, evutil_socket_t sv[2]);
      |                                                                      ^
```

When building with:

$ ./build.py -v --build-dir=/tmp/tritonbuild --enable-logging --enable-stats --enable-tracing --enable-metrics --endpoint=http --backend=identity

**Comment by 9cvele3** [comment]
> Did you build on Linux? I am seeing the following error:

Did not until now. I resolved that error (removed unneeded cast). Also, I had to sync with `main` branch to resolve some linker error unrelated to this PR. 

Linux build should be successful now with the command provided above.

**Comment by deadeyegoodwin** [comment]
The build command I gave you was missing an argument. Sorry about that. I see failures in building the new libevhtp:

$ ./build.py -v --build-dir=/tmp/tritonbuild --enable-logging --enable-stats --enable-tracing --enable-metrics --endpoint=http --backend=identity --repo-tag=thirdparty:<branch in your thirdparty repo>

You will also need to use the --github-organization to point to your github org that holds your forked thirdparty repo. This defaults to https://github.com/triton-inference-server but you need to set it so that it gets your modified thirdparty repo as part of the build

**Comment by 9cvele3** [comment]
I believe libevhtp build is fixed for Linux. 

However, I had some problems with `--github-organization` flag, it just didn't work for me. I had to change [this line in cmake](https://github.com/triton-inference-server/server/blob/78c52c7ae9bbbcaf6c67176664bebac42a45e930/build/CMakeLists.txt#L128) to be able to fetch my branch, and to test Linux build. Maybe I'm missing something ? I see that tag is properly passed here in cmake, but organization link is fixed to default one.

**Comment by 9cvele3** [comment]
>  I did get some error later in the process that I think is not related to this PR (`git clone of repo "identity_backend" at tag "main" failed `), and is probably caused with my cmake change.

Ignore this, I had some typo in my command. Linux build is successful.

**Comment by deadeyegoodwin** [comment]
Thanks for the updates. We are in the middle of 21.09 code freeze so I probably won't be able to look at this until next week.
I believe you are right that --github-organization is broken right now.

**Comment by 9cvele3** [comment]
Don't know how, but for me win build is successful without `winsock2.h`. 
But you are right, it should be there, so I added it.

**Pull request merged**